### ISSUE TITLE
ext\File-Find\t\taint.t failing on Windows

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#13458]
Migrated from [rt.perl.org#120711](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=120711) (status was 'resolved')

Searchable as RT120711$

**Comment by p5pRT** [comment]
### From @jkeenan

#### Created by @jkeenan


On 12/5/13 1&#8203;:08 PM\, Steve Hay posted on perl5\-porters the following report&#8203;:

\#\#\#\#\#
C&#8203;:\\Dev\\Git\\perl\\t>\.\\perl harness \-v \.\.\\ext\\File\-Find\\t\\taint\.t
\.\./ext/File\-Find/t/taint\.t \.\.
1\.\.54
ok 1 \- taint\.t found once
ok 2 \- taint\.t found once again
ok 3 \- able to mkdir&#8203;: for\_find
ok 4 \- successful chdir\(\) to for\_find
ok 5 \- able to mkdir&#8203;: fa
ok 6 \- able to mkdir&#8203;: fb
ok 7 \- able to create file&#8203;: fb\\fb\_ord
ok 8 \- able to mkdir&#8203;: fb\\fba
ok 9 \- able to create file&#8203;: fb\\fba\\fba\_ord
ok 10 \# skip Creating symlink
ok 11 \- able to create file&#8203;: fa\\fa\_ord
ok 12 \- able to mkdir&#8203;: fa\\faa
ok 13 \- able to create file&#8203;: fa\\faa\\faa\_ord
ok 14 \- able to mkdir&#8203;: fa\\fab
ok 15 \- able to create file&#8203;: fa\\fab\\fab\_ord
ok 16 \- able to mkdir&#8203;: fa\\fab\\faba
ok 17 \- able to create file&#8203;: fa\\fab\\faba\\faba\_ord
\# check untainting \(no follow\)
\# $File&#8203;::Find&#8203;::dir => 'fa'      $\_ => '\.'
ok 18 \- found \. for $\_\, as expected
\# $File&#8203;::Find&#8203;::dir => 'fa'      $\_ => 'fa\_ord'
ok 19 \- found fa\_ord for $\_\, as expected
\# $File&#8203;::Find&#8203;::dir => 'fa'      $\_ => 'faa'
ok 20 \- found faa for $\_\, as expected
\# $File&#8203;::Find&#8203;::dir => 'fa/faa'  $\_ => 'faa\_ord'
ok 21 \- found faa\_ord for $\_\, as expected
\# $File&#8203;::Find&#8203;::dir => 'fa'      $\_ => 'fab'
ok 22 \- found fab for $\_\, as expected
\# $File&#8203;::Find&#8203;::dir => 'fa/fab'  $\_ => 'fab\_ord'
ok 23 \- found fab\_ord for $\_\, as expectedinsecure cwd in find\(depth\)
at C&#8203;:\\Dev\\Git\\perl\\lib/File/Find\.pm line 620\.

\# $File&#8203;::Find&#8203;::dir => 'fa/fab'  $\_ => 'faba'
ok 24 \- found faba for $\_\, as expected
ok 25 \- Found all expected files
\# $File&#8203;::Find&#8203;::dir => 'fa'      $\_ => '\.'
ok 26 \- found \. for $\_\, as expected
\# $File&#8203;::Find&#8203;::dir => 'fa'      $\_ => 'fa\_ord'\# Looks like you planned
54 tests but ran 32\.

\# Looks like your test exited with 25 just after 32\.
ok 27 \- found fa\_ord for $\_\, as expected
\# $File&#8203;::Find&#8203;::dir => 'fa'      $\_ => 'faa'
ok 28 \- found faa for $\_\, as expected
\# $File&#8203;::Find&#8203;::dir => 'fa/faa'  $\_ => 'faa\_ord'
ok 29 \- found faa\_ord for $\_\, as expected
\# $File&#8203;::Find&#8203;::dir => 'fa'      $\_ => 'fab'
ok 30 \- found fab for $\_\, as expected
\# $File&#8203;::Find&#8203;::dir => 'fa/fab'  $\_ => 'fab\_ord'
ok 31 \- found fab\_ord for $\_\, as expected
\# $File&#8203;::Find&#8203;::dir => 'fa/fab'  $\_ => 'faba'
ok 32 \- found faba for $\_\, as expected
Dubious\, test returned 25 \(wstat 6400\, 0x1900\)
Failed 22/54 subtests
&nbsp;         \(less 1 skipped subtest&#8203;: 31 okay\)

Test Summary Report
\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
\.\./ext/File\-Find/t/taint\.t \(Wstat&#8203;: 6400 Tests&#8203;: 32 Failed&#8203;: 0\)
&nbsp;   Non\-zero exit status&#8203;: 25
&nbsp;   Parse errors&#8203;: Bad plan\.  You planned 54 tests but ran 32\.
Files=1\, Tests=32\,  0 wallclock secs \( 0\.02 usr \+  0\.00 sys =  0\.02 CPU\)
Result&#8203;: FAIL
\#\#\#\#\#

Since I have recently made a number of commits to blead
affecting ext/File\-Find\, I looked into this\.  The proximate
cause of this failure was the last of these commits&#8203;:

\#\#\#\#\#
commit 5ea8618bf5cf20c62d2ccca6aca10d97e0945b89
Author&#8203;: James E Keenan \<jkeenan@&#8203;cpan\.org>
Date&#8203;:   Wed Dec 4 03&#8203;:56&#8203;:13 2013 \+0100

&nbsp;     Add test for previously uncovered case\.

&nbsp;     Explicitly assigned $File&#8203;::Find&#8203;::untaint\_pattern\.

<details><summary>Inline Patch</summary>

```diff
diff --git a/ext/File-Find/t/taint.t b/ext/File-Find/t/taint.t
index 1675e6c..a32d10f 100644
--- a/ext/File-Find/t/taint.t
+++ b/ext/File-Find/t/taint.t
@@ -4,7 +4,7 @@ use Test::More;
  BEGIN {
      plan(
          ${^TAINT}
-        ? (tests => 45)
+        ? (tests => 54)
          : (skip_all => "A perl without taint support")
      );
  }
@@ -203,6 +203,30 @@ File::Find::find( {wanted => 

```
</details>
\&wanted_File_Dir_prune, untaint => 1,

&nbsp;  is\(scalar keys %Expect\_File\, 0\, 'Found all expected files'\);

\+\{
\+    %Expect\_File = \(File&#8203;::Spec\->curdir => 1\, file\_path\('fsl'\) =>
\+                    1\,file\_path\('fa\_ord'\) => 1\, file\_path\('fab'\) => 1\,
\+                    file\_path\('fab\_ord'\) => 1\, file\_path\('faba'\) => 1\,
\+                    file\_path\('faa'\) => 1\, file\_path\('faa\_ord'\) => 1\);
\+    delete $Expect\_File\{ file\_path\('fsl'\) \} unless $symlink\_exists;
\+    %Expect\_Name = \(\);
\+
\+    %Expect\_Dir = \( dir\_path\('fa'\) => 1\, dir\_path\('faa'\) => 1\,
\+                    dir\_path\('fab'\) => 1\, dir\_path\('faba'\) => 1\,
\+                    dir\_path\('fb'\) => 1\, dir\_path\('fba'\) => 1\);
\+
\+    delete @&#8203;Expect\_Dir\{ dir\_path\('fb'\)\, dir\_path\('fba'\) \} unless 
$symlink\_exists;
\+    local $File&#8203;::Find&#8203;::untaint\_pattern = qr|^\(\[@&#8203;\\w\./\\\-\+\]\+\)$|;
\+    File&#8203;::Find&#8203;::find\(
\+        \{
\+            wanted => \\&wanted\_File\_Dir\_prune\,
\+            untaint => 1\,
\+        \}\,
\+        topdir\('fa'\)
\+    \);
\+    is\(scalar keys %Expect\_File\, 0\,
\+        'Explicitly defined $File&#8203;::Find&#8203;::untaint\_pattern&#8203;: Found all 
expected files'\);
\+\}

&nbsp;  \# don't untaint at all\, should die
&nbsp;  %Expect\_File = \(\);
\#\#\#\#\#

This block was designed to exercise a branch in
lib/File/Find\.pm whose behavior was dependent on a
user\-defined $File&#8203;::Find&#8203;::untaint\_pattern\.  The block runs 9
unit tests&#8203;: 8 within wanted\_File\_Dir\_prune\(\)\, 1 being the
"Explicitly defined \.\.\." test above\.  I compared Steve's
report with subsequent smoke tests on Windows\.  Sometimes
the test file failed dubiously after 5 of the tests;
sometimes after 7\.

These failures may be the result of the interaction of
several factors\.

First\, the fact that the failures only emerged after I tried
to test $File&#8203;::Find&#8203;::untaint\_pattern may mean that that
package variable is conceptually wrong \-\- a true bug in
File&#8203;::Find itself\.  However\, it should be noted that quite a
few smoke reports had no problem with this test block\.  \(I
encountered no problems during development; otherwise I
would not have committed it to blead\.\)

Second\, this could be a Windows\-specific problem\, as the
subject of Steve's original post suggested\.  However\, since
the original report George Greer has sent me the results of
additional smoke tests on the smoke\-me branch in which I was
originally working\.  In this report\, generated on Linux
x86\_64\, there are repeated failures for runs with different
configurations\.  Example&#8203;:

\[perlio/en\_US\.utf8\] \-DDEBUGGING \-Duseithreads \-Dusemorebits 
\-Accflags=\-DPERL\_GLOBAL\_STRUCT
Inconsistent test results \(between TEST and harness\)&#8203;:
&nbsp;     \.\./t/re/fold\_grind\.t\.\.\.\.\.\.\.\. \.\.\.\.\.\.\.\.\.\.\.\.\. FAILED at test 10
&nbsp;     \.\./ext/File\-Find/t/taint\.t\.\.\.\.\. \.\.\.\.\.\.\.\.\.\.\.\.\. FAILED at test 3
&nbsp;     \.\./ext/File\-Find/t/find\.t\.\.\.\.\.\. \.\.\.\.\.\.\.\.\.\.\.\.\.\. FAILED at test 3

\(I don't really know what "inconsistent results between TEST
and harness" means\.\)

Third\, as I noted on p5p\, one of the things which I find
problematic about File&#8203;::Find's test suite \-\- but which I did
not attempt to remedy in my recent work \-\- is the fact that
it constructs all of the directories and files it uses to
test find\(\) and finddepth\(\) by hand underneath
'ext/File\-Find/t'\.  The test files make no use whatsoever of
File&#8203;::Temp or File&#8203;::Spec\->tmpdir\(\)\.  This bloats the code\,
imposes on the test author/maintainer the burden of manually
cleaning up all temporary directories and files\, and runs
the risk of clashes when testing in parallel \(\-j > 1\)\.

An immediate fix would be to revert the most recent commit\,
i\.e\.\, to remove the block of tests added to t/taint\.t\.
However\, reducing our test coverage is hardly an adequate
way of addressing test failure reports\.

Suggestions?

Thank you very much\.
Jim Keenan

<details><summary>Perl Info</summary>

```

Flags:
     category=library
     severity=low
     module=File::Find

Site configuration information for perl 5.18.0:

Configured by jimk at Sat May 18 12:07:56 EDT 2013.

Summary of my perl5 (revision 5 version 18 subversion 0) configuration:

   Platform:
     osname=darwin, osvers=8.11.0, archname=darwin-2level
     uname='darwin macintosh-9.local 8.11.0 darwin kernel version 
8.11.0: wed oct 10 18:26:00 pdt 2007; root:xnu-792.24.17~1release_ppc 
power macintosh powerpc '
     config_args='-des'
     hint=recommended, useposix=true, d_sigaction=define
     useithreads=undef, usemultiplicity=undef
     useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
     use64bitint=undef, use64bitall=undef, uselongdouble=undef
     usemymalloc=n, bincompat5005=undef
   Compiler:
     cc='cc', ccflags ='-fno-common -DPERL_DARWIN -fno-strict-aliasing 
-pipe -I/usr/local/include -I/opt/local/include',
     optimize='-O3',
     cppflags='-fno-common -DPERL_DARWIN -fno-strict-aliasing -pipe 
-I/usr/local/include -I/opt/local/include'
     ccversion='', gccversion='4.0.1 (Apple Computer, Inc. build 5250)', 
gccosandvers=''
     intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=4321
     d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
     ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', 
lseeksize=8
     alignbytes=8, prototype=define
   Linker and Libraries:
     ld='env MACOSX_DEPLOYMENT_TARGET=10.3 cc', ldflags =' 
-L/usr/local/lib -L/opt/local/lib'
     libpth=/usr/local/lib /opt/local/lib /usr/lib
     libs=-ldbm -ldl -lm -lc
     perllibs=-ldl -lm -lc
     libc=, so=dylib, useshrplib=false, libperl=libperl.a
     gnulibc_version=''
   Dynamic Linking:
     dlsrc=dl_dlopen.xs, dlext=bundle, d_dlsymun=undef, ccdlflags=' '
     cccdlflags=' ', lddlflags=' -bundle -undefined dynamic_lookup 
-L/usr/local/lib -L/opt/local/lib'

Locally applied patches:



@INC for perl 5.18.0:
     /usr/local/lib/perl5/site_perl/5.18.0/darwin-2level
     /usr/local/lib/perl5/site_perl/5.18.0
     /usr/local/lib/perl5/5.18.0/darwin-2level
     /usr/local/lib/perl5/5.18.0
     /usr/local/lib/perl5/site_perl/5.16.0
     /usr/local/lib/perl5/site_perl/5.14.2
     /usr/local/lib/perl5/site_perl/5.14.0
     /usr/local/lib/perl5/site_perl/5.12.0
     /usr/local/lib/perl5/site_perl/5.10.1
     /usr/local/lib/perl5/site_perl/5.10.0
     /usr/local/lib/perl5/site_perl/5.8.6
     /usr/local/lib/perl5/site_perl
     .


Environment for perl 5.18.0:
 
DYLD_LIBRARY_PATH=/Users/jimk/work/pseudoinstall/lib:/Users/jimk/gitwork/parrot/blib/lib
     HOME=/Users/jimk
     LANG=en_US.utf-8
     LANGUAGE (unset)
     LD_LIBRARY_PATH (unset)
     LOGDIR (unset)
 
PATH=/Users/jimk/perl5/perlbrew/bin:/usr/local/bin:/opt/local/bin:/opt/local/sbin:/usr/local/bin:/opt/local/bin:/opt/local/sbin:/bin:/sbin:/usr/bin:/usr/sbin:/Users/jimk/bin:/Users/jimk/bin/perl:/Users/jimk/bin/c:/Users/jimk/bin/shell:/sw/lib:/sw/bin:/Users/jimk/bin:/Users/jimk/bin/perl:/Users/jimk/bin/c:/Users/jimk/bin/shell:/sw/lib:/sw/bin
     PERLBREW_BASHRC_VERSION=0.59
     PERLBREW_HOME=/Users/jimk/.perlbrew
     PERLBREW_MANPATH=
     PERLBREW_PATH=/Users/jimk/perl5/perlbrew/bin
     PERLBREW_ROOT=/Users/jimk/perl5/perlbrew
     PERLBREW_VERSION=0.59
     PERL_BADLANG (unset)
     SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @bulk88

The taint\.t winds up calling a die in Find&#8203;::File\.


---
&nbsp;    unless \( $no\_chdir \) \{
&nbsp;        if \( \($check\_t\_cwd\) && \(\($untaint\) && \(is\_tainted\($cwd\) \)\) \) \{
&nbsp;        \( $cwd\_untainted \) = $cwd =~ m|$untaint\_pat|;
&nbsp;        unless \(defined $cwd\_untainted\) \{
&nbsp;            die "insecure cwd in find\(depth\)";\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<
&nbsp;        \}
&nbsp;        $check\_t\_cwd = 0;
&nbsp;        \}
&nbsp;        unless \(chdir $cwd\_untainted\) \{
&nbsp;        die "Can't cd to $cwd&#8203;: $\!\\n";
&nbsp;        \}
&nbsp;    \}


---
callstack\, taint\.t might be off by a line or 2 due to my hacking\, my blead is http&#8203;://perl5\.git\.perl\.org/perl\.git/commit/257dc59d7b864a6cf0ccc9179de1f3f0a797f4e0 \(yesterday/today's latest\)


---
Find\.pm\, line 620\, in File&#8203;::Find&#8203;::\_find\_opt
Find\.pm\, line 1081\, in File&#8203;::Find&#8203;::find
taint\.t\, line 220\, in main


---
from watch window\, var $cwd is "C&#8203;:/perl519/src/ext/File\-Find/for\_find"\, var $untaint\_pat is "\(?^&#8203;:^\(\[@&#8203;\\w\./\\\-\+\]\+\)$\)"\, var $cwd\_untainted is empty string or undef \(I can't tell\, its undef obviously for the die to execute\)\. $cwd didn't match $untaint\_pat regexp so the code did the die\. I should have given enough information for someone to figure out what to do\. I'm not sure what any of this code does so I won't analyze the problem further\.

\-\- 
bulk88 ~ bulk88 at hotmail\.com

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @jkeenan

On Fri Dec 06 09&#8203;:22&#8203;:40 2013\, bulk88 wrote&#8203;:
> The taint\.t winds up calling a die in Find&#8203;::File\.

Thanks for the lead\.  Hope to investigate more thoroughly this weekend\.

jimk

**Comment by p5pRT** [comment]
### From @jkeenan

On Fri Dec 06 09&#8203;:22&#8203;:40 2013\, bulk88 wrote&#8203;:
> The taint\.t winds up calling a die in Find&#8203;::File\.
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> unless \( $no\_chdir \) \{
>     if \( \($check\_t\_cwd\) && \(\($untaint\) && \(is\_tainted\($cwd\) \)\) \) \{
>         \( $cwd\_untainted \) = $cwd =~ m|$untaint\_pat|;
>         unless \(defined $cwd\_untainted\) \{
>             die "insecure cwd in
> find\(depth\)";\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<\<
>         \}
>         $check\_t\_cwd = 0;
>     \}
>     unless \(chdir $cwd\_untainted\) \{
>         die "Can't cd to $cwd&#8203;: $\!\\n";
>     \}
> \}
> 

My feeling is that the cause of the failures on Win32 has an OS\-specific aspect\.  In the relevant block of tests\, on Linux I do not enter the "unless \(defined $cwd\_untainted\)" branch\.

I loaded up the test file and the module with print statements&#8203;:

\#\#\#\#\#
sub \_find\_opt \{
\#\.\.\.

print STDERR "ZZZ&#8203;:"\, "\\n";

&nbsp;    unless \( $no\_chdir \) \{
print STDERR "AAA&#8203;:"\, "\\n";
&nbsp;        if \( \($check\_t\_cwd\) && \(\($untaint\) && \(is\_tainted\($cwd\) \)\) \) \{
print STDERR "BBB&#8203;:"\, "\\n";
&nbsp;        \( $cwd\_untainted \) = $cwd =~ m|$untaint\_pat|;
&nbsp;        unless \(defined $cwd\_untainted\) \{
print STDERR "CCC&#8203;:"\, "\\n";
&nbsp;            die "insecure cwd in find\(depth\)";
&nbsp;        \}
&nbsp;        $check\_t\_cwd = 0;
&nbsp;        \}
&nbsp;        unless \(chdir $cwd\_untainted\) \{
&nbsp;        die "Can't cd to $cwd&#8203;: $\!\\n";
&nbsp;        \}
&nbsp;    \}
&nbsp;    \}
\}
\#\#\#\#\#

For the block in the test file which tests $File&#8203;::Find&#8203;::untaint\_pattern\, I hit ZZZ\, AAA and BBB\, but not CCC\.  \(I do hit CCC in other test blocks\.\)  I got the same results on Darwin/PPC\.

So it's not clear how we should proceed\.

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
### From @bulk88

On Fri Dec 06 17&#8203;:53&#8203;:55 2013\, jkeenan wrote&#8203;: 
> For the block in the test file which tests
> $File&#8203;::Find&#8203;::untaint\_pattern\, I hit ZZZ\, AAA and BBB\, but not CCC\.  \(I
> do hit CCC in other test blocks\.\)  I got the same results on
> Darwin/PPC\.
> 
> So it's not clear how we should proceed\.

If you hit BBB\, and never CCC\, on Unix\, then it has something to do with the path and the regexp\. 

>     \( $cwd\_untainted \) = $cwd =~ m|$untaint\_pat|;

In all likelyhood this is a probably a bug in the \.t or the \.pm where a regexp only understands Unix paths and not Win32 paths\.

To die\, the match must fail\, on my win32 system\, $cwd was 'C&#8203;:/perl519/src/ext/File\-Find/for\_find'\, $untaint\_pat is '\(?^&#8203;:^\(\[@&#8203;\\w\./\\\-\+\]\+\)$\)' \(' intentional\, that $ on the end can't go through "\)\. To make the test pass\, first question is why did 'C&#8203;:/perl519/src/ext/File\-Find/for\_find' not match the regexp '\(?^&#8203;:^\(\[@&#8203;\\w\./\\\-\+\]\+\)$\)'? \(I won't pretend to understand the code written by hungry moggy on a keyboard\) Next question is\, where did $untaint\_pat come from \(test supplied\, or \.pm supplied\)\, and should it be '\(?^&#8203;:^\(\[@&#8203;\\w\./\\\-\+\]\+\)$\)' or something else? The third question is\, is 'C&#8203;:/perl519/src/ext/File\-Find/for\_find' \($cwd\) correct or should it be a just a file name or relative path or something \(and where did it come from? C&#8203;:/perl519/src is my interp source code dir\.

If your can't fix this with deep understanding \(which I don't have any of\) of File\-Find\, atleast post what $cwd and $untaint\_pat are at point BBB for the record\, on your Linux\, at the test that fails on Win32 which is


---
&nbsp;    local $File&#8203;::Find&#8203;::untaint\_pattern = qr|^\(\[@&#8203;\\w\./\\\-\+\]\+\)$|;
&nbsp;    File&#8203;::Find&#8203;::find\(
&nbsp;        \{
&nbsp;            wanted => \\&wanted\_File\_Dir\_prune\,
&nbsp;            untaint => 1\,
&nbsp;        \}\,
&nbsp;        topdir\('fa'\)
&nbsp;    \);
&nbsp;    is\(scalar keys %Expect\_File\, 0\,
&nbsp;        'Explicitly defined $File&#8203;::Find&#8203;::untaint\_pattern&#8203;: Found all expected files'\);


---
\-\- 
bulk88 ~ bulk88 at hotmail\.com

**Comment by p5pRT** [comment]
### From @jkeenan

On Fri Dec 06 13&#8203;:41&#8203;:53 2013\, jkeenan wrote&#8203;:
> On Fri Dec 06 09&#8203;:22&#8203;:40 2013\, bulk88 wrote&#8203;:
> > The taint\.t winds up calling a die in Find&#8203;::File\.
> 
> Thanks for the lead\.  Hope to investigate more thoroughly this weekend\.
> 
> jimk


I have created this branch for smoke\-testing&#8203;:

smoke\-me/jkeenan/120711\-file\-find\-tests

This branch uses File&#8203;::Temp\->newdir\(\) to create a true tempdir for the top level of testing\.

Smoking on Win32 and VMS would be very appreciated\.

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
### From @bulk88

On Fri Dec 06 19&#8203;:33&#8203;:11 2013\, jkeenan wrote&#8203;:
> 
> 
> I have created this branch for smoke\-testing&#8203;:
> 
> smoke\-me/jkeenan/120711\-file\-find\-tests
> 
> This branch uses File&#8203;::Temp\->newdir\(\) to create a true tempdir for the
> top level of testing\.
> 
> Smoking on Win32 and VMS would be very appreciated\.
> 
> Thank you very much\.
> Jim Keenan

failed\, same Find&#8203;::File die as before\, new warning from File&#8203;::Temp


---
1\.\.52
ok 1 \- taint\.t found once
ok 2 \- taint\.t found once again
ok 3 \- able to mkdir&#8203;: fa
ok 4 \- able to mkdir&#8203;: fb
ok 5 \- able to create file&#8203;: fb\\fb\_ord
ok 6 \- able to mkdir&#8203;: fb\\fba
ok 7 \- able to create file&#8203;: fb\\fba\\fba\_ord
ok 8 \# skip Creating symlink
ok 9 \- able to create file&#8203;: fa\\fa\_ord
ok 10 \- able to mkdir&#8203;: fa\\faa
ok 11 \- able to create file&#8203;: fa\\faa\\faa\_ord
ok 12 \- able to mkdir&#8203;: fa\\fab
ok 13 \- able to create file&#8203;: fa\\fab\\fab\_ord
ok 14 \- able to mkdir&#8203;: fa\\fab\\faba
ok 15 \- able to create file&#8203;: fa\\fab\\faba\\faba\_ord
\# check untainting \(no follow\)
\# $File&#8203;::Find&#8203;::dir => 'fa'    $\_ => '\.'
ok 16 \- found \. for $\_\, as expected
\# $File&#8203;::Find&#8203;::dir => 'fa'    $\_ => 'fa\_ord'
ok 17 \- found fa\_ord for $\_\, as expected
\# $File&#8203;::Find&#8203;::dir => 'fa'    $\_ => 'faa'
ok 18 \- found faa for $\_\, as expected
\# $File&#8203;::Find&#8203;::dir => 'fa/faa'    $\_ => 'faa\_ord'
ok 19 \- found faa\_ord for $\_\, as expected
\# $File&#8203;::Find&#8203;::dir => 'fa'    $\_ => 'fab'
ok 20 \- found fab for $\_\, as expected
\# $File&#8203;::Find&#8203;::dir => 'fa/fab'    $\_ => 'fab\_ord'
ok 21 \- found fab\_ord for $\_\, as expected
\# $File&#8203;::Find&#8203;::dir => 'fa/fab'    $\_ => 'faba'
ok 22 \- found faba for $\_\, as expected
ok 23 \- Found all expected files
\# $File&#8203;::Find&#8203;::dir => 'fa'    $\_ => '\.'
ok 24 \- found \. for $\_\, as expected
\# $File&#8203;::Find&#8203;::dir => 'fa'    $\_ => 'fa\_ord'
ok 25 \- found fa\_ord for $\_\, as expected
\# $File&#8203;::Find&#8203;::dir => 'fa'    $\_ => 'faa'
ok 26 \- found faa for $\_\, as expected
\# $File&#8203;::Find&#8203;::dir => 'fa/faa'    $\_ => 'faa\_ord'
ok 27 \- found faa\_ord for $\_\, as expected
\# $File&#8203;::Find&#8203;::dir => 'fa'    $\_ => 'fab'
ok 28 \- found fab for $\_\, as expected
\# $File&#8203;::Find&#8203;::dir => 'fa/fab'    $\_ => 'fab\_ord'
ok 29 \- found fab\_ord for $\_\, as expected
\# $File&#8203;::Find&#8203;::dir => 'fa/fab'    $\_ => 'faba'
ok 30 \- found faba for $\_\, as expected
insecure cwd in find\(depth\) at C&#8203;:\\perl519\\src\\lib/File/Find\.pm line 620\.
cannot remove path when cwd is C&#8203;:/WINDOWS/TEMP/XqF\_GVZ8fJ/fa/fab for C&#8203;:/WINDOWS/TEMP/XqF\_GVZ8fJ&#8203;:  at C&#8203;:\\perl519\\src\\lib/File/Temp\.pm line 1616\.
\# Looks like you planned 52 tests but ran 30\.
\# Looks like your test exited with 25 just after 30\.


---
\-\- 
bulk88 ~ bulk88 at hotmail\.com

**Comment by p5pRT** [comment]
### From @jkeenan

On Fri Dec 06 22&#8203;:27&#8203;:18 2013\, bulk88 wrote&#8203;:
> On Fri Dec 06 19&#8203;:33&#8203;:11 2013\, jkeenan wrote&#8203;:
> >
> >
> > I have created this branch for smoke\-testing&#8203;:
> >
> > smoke\-me/jkeenan/120711\-file\-find\-tests
> >
> > This branch uses File&#8203;::Temp\->newdir\(\) to create a true tempdir for
> > the
> > top level of testing\.
> >
> > Smoking on Win32 and VMS would be very appreciated\.
> >
> > Thank you very much\.
> > Jim Keenan
> 
> failed\, same Find&#8203;::File die as before\, new warning from File&#8203;::Temp
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> 1\.\.52
> ok 1 \- taint\.t found once
> ok 2 \- taint\.t found once again
> ok 3 \- able to mkdir&#8203;: fa
> ok 4 \- able to mkdir&#8203;: fb
> ok 5 \- able to create file&#8203;: fb\\fb\_ord
> ok 6 \- able to mkdir&#8203;: fb\\fba
> ok 7 \- able to create file&#8203;: fb\\fba\\fba\_ord
> ok 8 \# skip Creating symlink
> ok 9 \- able to create file&#8203;: fa\\fa\_ord
> ok 10 \- able to mkdir&#8203;: fa\\faa
> ok 11 \- able to create file&#8203;: fa\\faa\\faa\_ord
> ok 12 \- able to mkdir&#8203;: fa\\fab
> ok 13 \- able to create file&#8203;: fa\\fab\\fab\_ord
> ok 14 \- able to mkdir&#8203;: fa\\fab\\faba
> ok 15 \- able to create file&#8203;: fa\\fab\\faba\\faba\_ord
> \# check untainting \(no follow\)
> \# $File&#8203;::Find&#8203;::dir => 'fa'      $\_ => '\.'
> ok 16 \- found \. for $\_\, as expected
> \# $File&#8203;::Find&#8203;::dir => 'fa'      $\_ => 'fa\_ord'
> ok 17 \- found fa\_ord for $\_\, as expected
> \# $File&#8203;::Find&#8203;::dir => 'fa'      $\_ => 'faa'
> ok 18 \- found faa for $\_\, as expected
> \# $File&#8203;::Find&#8203;::dir => 'fa/faa'  $\_ => 'faa\_ord'
> ok 19 \- found faa\_ord for $\_\, as expected
> \# $File&#8203;::Find&#8203;::dir => 'fa'      $\_ => 'fab'
> ok 20 \- found fab for $\_\, as expected
> \# $File&#8203;::Find&#8203;::dir => 'fa/fab'  $\_ => 'fab\_ord'
> ok 21 \- found fab\_ord for $\_\, as expected
> \# $File&#8203;::Find&#8203;::dir => 'fa/fab'  $\_ => 'faba'
> ok 22 \- found faba for $\_\, as expected
> ok 23 \- Found all expected files
> \# $File&#8203;::Find&#8203;::dir => 'fa'      $\_ => '\.'
> ok 24 \- found \. for $\_\, as expected
> \# $File&#8203;::Find&#8203;::dir => 'fa'      $\_ => 'fa\_ord'
> ok 25 \- found fa\_ord for $\_\, as expected
> \# $File&#8203;::Find&#8203;::dir => 'fa'      $\_ => 'faa'
> ok 26 \- found faa for $\_\, as expected
> \# $File&#8203;::Find&#8203;::dir => 'fa/faa'  $\_ => 'faa\_ord'
> ok 27 \- found faa\_ord for $\_\, as expected
> \# $File&#8203;::Find&#8203;::dir => 'fa'      $\_ => 'fab'
> ok 28 \- found fab for $\_\, as expected
> \# $File&#8203;::Find&#8203;::dir => 'fa/fab'  $\_ => 'fab\_ord'
> ok 29 \- found fab\_ord for $\_\, as expected
> \# $File&#8203;::Find&#8203;::dir => 'fa/fab'  $\_ => 'faba'
> ok 30 \- found faba for $\_\, as expected
> insecure cwd in find\(depth\) at C&#8203;:\\perl519\\src\\lib/File/Find\.pm line
> 620\.
> cannot remove path when cwd is C&#8203;:/WINDOWS/TEMP/XqF\_GVZ8fJ/fa/fab for
> C&#8203;:/WINDOWS/TEMP/XqF\_GVZ8fJ&#8203;:  at C&#8203;:\\perl519\\src\\lib/File/Temp\.pm line
> 1616\.
> \# Looks like you planned 52 tests but ran 30\.
> \# Looks like your test exited with 25 just after 30\.
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

Since in smoke test reports I'm getting the untaint\_pattern failures on both Win32 \(always\) and Linux \(sometimes\)\, I'm removing the block that I added\.

I stumbled upon this only because I was attempting to improve File\-Find's test coverage\.  For the time being I'll focus on trying to find ways to do that that don't run into tainting problems\.

In the branch I created\, I'll TODO that test block\, but I'm still interested in seeing whether $File&#8203;::Temp\->newdir\(\) helps to alleviate problems in parallel testing\.

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
### From @jkeenan

On Sat Dec 07 05&#8203;:56&#8203;:26 2013\, jkeenan wrote&#8203;:

> 
> Since in smoke test reports I'm getting the untaint\_pattern failures
> on both Win32 \(always\) and Linux \(sometimes\)\, I'm removing the block
> that I added\.
> 

Done in a5e668419a0ce5ec48b3586fb224eed5ace5d64c\.

> I stumbled upon this only because I was attempting to improve File\-
> Find's test coverage\.  For the time being I'll focus on trying to find
> ways to do that that don't run into tainting problems\.
> 

A few additional tests were committed in 1c6ce212d37a2a17424675e94afb035a9a446f9f\.

> In the branch I created\, I'll TODO that test block\, but I'm still
> interested in seeing whether $File&#8203;::Temp\->newdir\(\) helps to alleviate
> problems in parallel testing\.
> 

$File&#8203;::Temp\->newdir\(\) so far has not tested cleanly\.  If I can figure out why\, I'll open a new ticket\.

Since the original complaint has been addressed\, I'm closing this ticket\.

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
@jkeenan - Status changed from 'open' to 'resolved'
